shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,773 @@
+  isSubsetOf(other) {
+    check(other, PermissionSet);
+    for (let ii = 0; ii < this.array.length; ++ii) {
+      const mine = this.array[ii];
+      const yours = other.array[ii] || false;
+      if (mine != yours) {
+        return false;
+      }
     }
+
+    return true;
+  }
+
+  // Methods for mutating a PermissionSet by combining it with another PermissionSet.
+  // These return a boolean indicating whether the operation had any effect.
+
+  add(other) {
+    check(other, PermissionSet);
+    let changed = false;
+    for (let ii = 0; ii < other.array.length; ++ii) {
+      const old = this.array[ii];
+      this.array[ii] = this.array[ii] || other.array[ii];
+      if (old != this.array[ii]) {
+        changed = true;
+      }
+    }
+
+    return changed;
+  }
+
+  remove(other) {
+    check(other, PermissionSet);
+    let changed = false;
+    for (let ii = 0; ii < other.array.length && ii < this.array.length; ++ii) {
+      const old = this.array[ii];
+      this.array[ii] = this.array[ii] && !other.array[ii];
+      if (old != this.array[ii]) {
+        changed = true;
+      }
+    }
+
+    return changed;
   }
-  return changed;
-}
 
-PermissionSet.prototype.intersect = function(other) {
-  var changed = false;
-  for (var ii = 0; ii < this.array.length; ++ii) {
-    var old = this.array[ii];
-    this.array[ii] = this.array[ii] && other.array[ii];
-    if (old != this.array[ii]) {
-      changed = true;
+  intersect(other) {
+    check(other, PermissionSet);
+    let changed = false;
+    for (let ii = 0; ii < this.array.length; ++ii) {
+      const old = this.array[ii];
+      this.array[ii] = this.array[ii] && other.array[ii];
+      if (old != this.array[ii]) {
+        changed = true;
+      }
     }
+
+    return changed;
   }
-  return changed;
 }
 
-function roleAssignmentPermissions(roleAssignment, viewInfo) {
-  var result = new PermissionSet([]);
+class Clause {
+  // A conjunction of permissions for identities on grains.
+
+  constructor() {
+    this.identityPermissions = {};
+    // Two-level map. Maps a pair of a grain ID and an identity ID to a PermissionSet.
+  }
 
-  if (!roleAssignment || "none" in roleAssignment) {
-    if (viewInfo.roles) {
-      for (var ii = 0; ii < viewInfo.roles.length; ++ii) {
-        var roleDef = viewInfo.roles[ii];
-        if (roleDef.default) {
-          result = new PermissionSet(roleDef.permissions);
-          break;
+  hash() {
+    // It is often useful to form sets of clauses. This method returns a hash that is suitable to
+    // be used as a key for such a set.
+    //
+    // TODO(soon): Investigate other approaches. We do not need this hash to be cryptographically
+    //   secure; we just need it to avoid accidental collisions. Maybe we should instead represent
+    //   sets of clauses as ordered maps, or maybe as proper hash maps with bucketing?
+    const hasher = Crypto.createHash("sha256");
+    for (let grainId in this.identityPermissions) {
```

Comment:
Do you need to make sure these are in a canonical order? Note that iterating over keys in a JS object iterates in the order in which the keys were added, not lexical order.